### PR TITLE
Add `jsx-import` rule

### LINKED ISSUE #155: Rule proposal: `jsx-import`
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/155)

Idea from: https://github.com/sindresorhus/refined-github/pull/1089/files#r168391471

When JSX is detected in the source, require the code to import one of the following imports:

- `import React from 'react';`
- `import preact from 'preact';`
- `import {h} from 'preact';`
- `import {h} from 'dom-chef';`
- … (We can fill this list out from need)

And if no JSX is is detect, disallow these imports.

Need suggestions for a better rule name.

Thoughts on the rule?

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/155).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by kevva**
I think this'll need some work regarding which pragmas we allow and how the user can adjust them.

Fixes #155.

### CODE REVIEW

File: rules/jsx-import.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+				const isDeclaration = isImportDeclaration(pragma.defs[0].node);
+				const packages = pragmas[pragma.name];
+				const hasPkg = hasPackage(node, packages);
+
+				if (!isJsx) {
+					context.report({
+						node,
+						message: `\`${pragma.name}\` shouldn't be imported when not using JSX`
+					});
+				}
```
Comment (kevva) [inline rules/jsx-import.js]: Should we accept use of `React.cloneElement()`? We could check if the variable references are empty, but looking at [this](https://github.com/eslint/eslint/blob/master/lib/rules/no-unused-vars.js) it sounds easier than it is.

File: test/jsx-import.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		`import preact from 'preact'; let Foo; <Foo/>;`,
+		`import preact, {render} from 'preact'; let Foo; <Foo/>;`,
+		`import {h, render} from 'preact'; let Foo; <Foo/>;`,
+		`import {h} from 'dom-chef'; let Foo; <Foo/>;`,
+		`import React from 'react'; import bar from 'bar'; let Foo; <Foo/>;`,
+		`const React = require('react'); let Foo; <Foo/>;`,
+		`const preact = require('preact'); let Foo; <Foo/>;`,
+		`const {h} = require('preact'); let Foo; <Foo/>;`,
+		`const {h} = require('dom-chef'); let Foo; <Foo/>;`,
+		`const {h} = require('dom-chef'); const bar = require('bar'); let Foo; <Foo/>;`
```
Comment (sindresorhus) [inline test/jsx-import.js]: Can you add some tests where the JSX is nested in a couple of functions?

File: rules/jsx-import.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	return variables.reverse();
+};
+
+const create = context => {
+	let isJsx = false;
+	let pragma;
+
+	return {
+		Program: () => {
+			for (const x of Object.keys(pragmas)) {
```
Comment (sindresorhus) [inline rules/jsx-import.js]: `x` => `key`

**Comment by sindresorhus** [comment]
@kevva Still interested in finishing this?

**Comment by sindresorhus** [comment]
Closing for now. Happy to reopen if you ever get to this, @kevva ;)

**Pull request closed without merging**